Feature/8537 site creation segments ui

malinajirka opened the pull request:
Fixes #8537 

This PR is based on the branch from #8532 and therefor it's change set is big -> I recommend reviewing it after the #8532 is merged.

Adds UI to the SiteCreation - segments screen. 

Error screen - We don't have a design for the error screen yet, therefor this PR introduces just a dummy screen.

To test - change `NEW_SITE_CREATION_ENABLED` build time variable to 'true' in build.gradle
1. Go to My Site 
2. Click on Switch Site
3. Click on Plus sign icon
4. Select Create WordPress.com site
5. Test the screen - progress, list, error with retry (+rotate your device and test tablet layout)
Notes: 
- Selecting a Segment doesn't have any effect yet.
- you need to modify code in order to test error state since we are using a mocked api -> for example add `event.error = FetchSegmentsError(GENERIC_ERROR,"test")` into `FetchSegmentsUseCase.onSiteCategoriesFetched()`

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentViewHolder.kt:
```diff
@@ -0,0 +1,39 @@
+import org.wordpress.android.util.image.ImageManager
+import org.wordpress.android.util.image.ImageType.IMAGE
+
+sealed class NewSiteCreationSegmentViewHolder(internal val parent: ViewGroup, @LayoutRes layout: Int) :
```
I assume you're planning to add more view holders that's why you have a `sealed class`. Is that right?

malinajirka replied:
I'm not really planning to add more, but I feel like this is a good pattern and it's easily open for extension when we need to add more types, yet it doesn't make things complex. So it felt like a "good deal".

But tbh maybe we should add even the header into the recycler view so the whole screen is scrollable - wdyt?

malinajirka replied:
Maybe it's not such a good idea to put the header into the recycler view as we want to be able to show the header + progress bar -> we'd have to create another ListItemType for progress, but it'd a bit complicated to make it match the rest of the screen.

So question is whether you think I should remove the sealed class when I don't plan to create another item type at the moment.

oguzkocer replied:
> So question is whether you think I should remove the sealed class when I don't plan to create another item type at the moment

I think that's best for now. It's easy enough to refactor when we need it.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+) : Adapter<NewSiteCreationSegmentViewHolder>() {
+    private val items = mutableListOf<VerticalSegmentModel>()
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SegmentViewHolder {
```
Should this return `NewSiteCreationSegmentViewHolder` instead?

malinajirka replied:
👍 Good catch!

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+        val oldItems: List<VerticalSegmentModel>,
+        val newItems: List<VerticalSegmentModel>
+    ) :
+            DiffUtil.Callback() {
```
Nitpick: Can we move this to the above line? Sometimes the auto-formatter does this 😠

malinajirka replied:
👍

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentViewHolder.kt:
```diff
@@ -0,0 +1,39 @@
+
+sealed class NewSiteCreationSegmentViewHolder(internal val parent: ViewGroup, @LayoutRes layout: Int) :
+        RecyclerView.ViewHolder(LayoutInflater.from(parent.context).inflate(layout, parent, false)) {
+    abstract fun onBind(model: VerticalSegmentModel, isLast: Boolean)
```
This is completely fine for this feature, especially right now when we don't have any details in the model anyway. However, I'd really love to see an adapter item that we map to instead of using the model directly if the model ever becomes complicated. Just something to keep in mind for future.

malinajirka replied:
I was thinking about creating something like `ItemUiState` with visibility states + onClickLambda + data. Is this something you have in mind?

malinajirka replied:
It could either have `VerticalSegmentModel` in an internal field and expose only required data or it could have separate fields and we'd map the data in the VM.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentViewHolder.kt:
```diff
@@ -0,0 +1,39 @@
+        override fun onBind(model: VerticalSegmentModel, isLast: Boolean) {
+            title.text = model.title
+            subtitle.text = model.subtitle
+            imageManager.load(icon, IMAGE, StringUtils.notNullStr(model.iconUrl))
```
```suggestion
            if(!model.iconUrl.isBlank()) {
                imageManager.load(icon, IMAGE, model.iconUrl)
            } else {
                imageManager.cancelRequestAndClearImageView(icon)
            }
```
Do we need to do something like this ^? The `ImageManager.load` method says that the empty values will be used to load the placeholder, but it doesn't actually cancel the previous request, does it? If it does, it kind of makes me question why we use the `cancelRequestAndClearImageView` in some cases that we do in other pages.

At the very least, I think we can get rid of the `StringUtils.notNullStr` since it's not a null-able value. I can see an argument for keeping it in case for future changes, but that's a slippery slope and I feel like if it's best to keep the code clean.

malinajirka replied:
Good question!
- AFAIK invoking a load method on the same ImageView automatically cancel the previous request
- We use `cancelRequestAndClearImageView` for cases where for example the view for that particular item is completely hidden - as we don't want to show any placeholder + we want to free the memory --->> we could theoretically call load with "", but that would make the code unreadable in my opinion


>At the very least, I think we can get rid of the StringUtils.notNullStr since it's not a null-able value. 
I agree -> I was developing it against the previous branch where the `iconUrl` was nullable.

malinajirka replied:
👍

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    fun update(newItems: List<VerticalSegmentModel>) {
+        val diffResult = DiffUtil.calculateDiff(SegmentsDiffUtils(items.toList(), newItems))
```
Generally speaking, it's not really a good idea to calculate the differences in the main thread. It should be fine as is for now, but we might eventually have to make that change. This is another case where `PagedListAdapter` really shines 😀

malinajirka replied:
Yes, I agree. However I'd improve that when we have time as it shouldn't have any effect for this particular screen. But thanks for the comment!

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+import org.wordpress.android.util.image.ImageManager
+
+class NewSiteCreationSegmentsAdapter(
+    private val onItemTapped: (VerticalSegmentModel) -> Unit = { },
```
This is actually a good case for why we should not use the `VerticalModel` directly and have an adapter item. For the item tapped action, the adapter plays the role of a middle man where it doesn't contribute to anything. If we had the adapter item, we could have separate `onItemTapped` actions for each adapter item and the Adapter or the ViewHolder wouldn't actually need to know anything about the `VerticalModel`.

Again, I am not actually asking you to make this change. You mentioned you were curious about PR reviews and I just wanted to be more elaborate on my thoughts when I am reviewing this PR in case you find it useful. I normally don't bring up all these things, so if you don't find it useful, I am happy to ignore the things we don't need yet.

malinajirka replied:
I'll be happy to make this change and I enjoy these comments as they make me think + I learn from them.
I'll wait for you comment https://github.com/wordpress-mobile/WordPress-Android/pull/8547#discussion_r231315975 before I make the change, so I'm sure we are on the same page.

oguzkocer commented:
@malinajirka I followed up on all your comments, but haven't done a second review of the code yet, since there will be some changes. Let me know if I missed anything!

malinajirka commented:
Thanks @oguzkocer!!

Since one of the (new) requirements is to make the header scroll with the content I decided to put both header and progress views into the RecyclerView. I also introduced `ItemUiState` as we discussed. I'm sorry for the number of changes and the size of the PR now :(( (I also had to fix tests).

>The reason I separated the data from the onClick events, even though it's specific to each row, is to make it easier to compare the data for areContentsTheSame diff callback. 
I added the onItemTapped attribute as a `var` in to the data class instead of putting into the constructor - so it isn't part of the automatically generated `equals, toString, hashCode` methods.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentViewHolder.kt:
```diff
@@ -0,0 +1,59 @@
+            subtitle.text = uiState.subtitle
+            imageManager.load(icon, IMAGE, uiState.iconUrl)
+            container.setOnClickListener { uiState.onItemTapped }
+            divider.visibility = if (uiState.showDivider) View.VISIBLE else View.GONE
```
I don't think there is anything wrong doing it like this, but I believe you can use the item decorator for the divider. This is just an fyi, if you didn't know, no changes necessary.

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/segments/NewSiteCreationSegmentsViewModel.kt:
```diff
@@ -73,38 +93,86 @@ class NewSiteCreationSegmentsViewModel
+            val iconUrl: String,
+            val showDivider: Boolean
+        ) : ItemUiState() {
+            lateinit var onItemTapped: (Long) -> Unit
```
I can't say I'm a big fan either -> however I didn't find any better way how to exclude an attribute from the generated methods.
